### ISSUE TITLE
Can you broadcast to a GID other than 111-111-1111?

### ISSUE THREAD

**Issue by jeffjvick** [gotenna/PublicSDK#38]
The sendBroadcast command in the documentation shows a destination GID (that it says can't be changed) however when you try to use the command it seems that "option" has been removed, not GID is needed to be set, it uses the default shout GID.

It would seem you can only broadcast to 111-111-1111 (and the emergency GID) so this is more of a feature request so could this be added?  Some of the use cases for the SDK I'm considering would involve sending out some type of beacon to other people using my app but I don't want everyone receiving those broadcasts as they would be annoying.  

Or does broadcasting to 111-111-1111 with a different SDK token mean that only people using my app would see that broadcast and not other people on different apps with different tokens?  The official goTenna app see's broadcasts from my app's token but I don't know if this is expected behavior or a bug.

It would be helpful to be able to broadcast messages to just users of a specific app.

**Comment by tcolligan-ap** [comment]
Hi Jeff,

Other apps being able to see your shouts is definitely a bug! Ideally all messages should be "sandboxed" by your sdk token, meaning only other apps with that have your sdk token should be able to receive those messages.

Being about to have something like your own custom shouts channel that you can dynamically define is something that we are currently looking into implementing, but will more than likely not make it into the next release. The bug that you mentioned however, is planned to be fixed in the next release.

In the meantime, you could use the groups functionality to do something similar to you own private shouts, the downside being, you need to know the GID of everyone beforehand in-order to invite them to a group, so it is obviously not as flexible as shouts.

**Comment by jeffjvick** [comment]
@tcolligan-ap Thanks for the response.  Good to hear the shouts / token issue is a bug.  

If (once this is fixed) other apps (including the official one) can see my app's shouts to GID 111-111-1111 that is probably a good enough start to test some ideas I have.  Ability to make your own shout 'channels' would likely be a good later feature as one channel could be used for discovery operations and then other channels could be used to 'subscribe' to some kind of service.   

Does sending a message to a group just create X number of 1:1 messages internally on the device or does it do something more efficient like shout to a list of GIDs?

**Comment by tcolligan-ap** [comment]
@jeffjvick It is more efficient than just sending a number of 1:1 messages internally.

The goTenna can actually have a number of GIDs assigned to it at once (with some limitations).

In a normal scenario, your private GID is the only one registered to your goTenna. However, when you create or are invited to a group, that group has its own unique GID, which is then registered onto your goTenna along with your private GID. This essentially allows your goTenna to receive any messages that are addressed to either of those GIDs.

In the initial group invite phase, everyone in the group does get a separate 1:1 message, which is the invite to the group with all the relevant info about the group. After that, each individual registers the group GID onto their goTenna. They can then send messages to the Group GID as if it were an individual, when actually it is a collection of individuals.

There is a bit more going on under the hood beyond that, which I haven't really gone into much detail on, but from a high level that is essentially what is happening.

**Comment by jeffjvick** [comment]
@tcolligan-ap Thanks for the insight on what is going on under the hood with regards to group messages.  I like the way this is structured.

In the SDK docs it says that a group cannot exceed 10 members.  Since each device in a group shares the same group GID and for a group message only one message to that GID is sent, it doesn't seem like the sending / receiving would have a technical limit of 10 (or really any limit).  

Is the 10 limit just because the initial 1:1 messaging to create the group becomes too slow as the number gets bigger than 10?  And might there be a way to add additional members to your group?  I don't see an SDK command for that currently.

**Comment by tcolligan-ap** [comment]
@jeffjvick You are correct, with the way the whole system is setup, in theory the limit of 10 is somewhat arbitrary, but there are various reasons for it.

One of them is just from a product standpoint, a group of 10 fits the vast majority of current use cases for goTenna. Shouts is generally seen as a viable alternative if you need something more dynamic.

Another reason is, like you said, the initial group creation would take quite a while if we let that number scale up much further, and would have the potential of flooding the network with invite messages. If a delay was put in to prevent network flooding, it would become an even more unbearably slow process.

Finally just for general QA and support sanity, setting up 10 Mobile Phones with 10 goTenna devices to make sure that the whole invitation and group messaging process works properly at that scale is already not a quick or simple test to get setup to be sure, expanding beyond that just makes it more complicated. It can be done for sure, but there is always a balance between demand, cost, and benefit.

Adding additional members post group creation runs into similar issues, in that you would have to send an additional private invite to that member, and also notify existing members that an additional member is added with meta data info on who they are, otherwise it would appear as if some random person just joined your group.

This is also partially an artifact of the Group Acks feature from the V1 goTenna, which allowed you to know which members in the group actually received your message or not. Since this is an off the grid system with limited bandwidth and storage capacity, waiting to collect acks from a certain number of group members had to be taken into consideration as well.

Again, these are all surmountable technical challenges, it is just very much a decision on what types of features we want to focus on and add value to at the moment and in the long run. Hope that explanation helped!
